File: src/core/loginItem.ts
import { existsSync, mkdirSync, unlinkSync, writeFileSync } from "fs";
import { homedir } from "os";
import { dirname, join } from "path";

const AGENTS = join(homedir(), "Library/LaunchAgents");
const LOGIN_LABEL = "com.limit-starter.login";
const LOGIN_PLIST = join(AGENTS, `${LOGIN_LABEL}.plist`);
/** The launchd daemon from the CLI-only version; the menu bar app replaces it. */
const LEGACY_LABEL = "com.limit-starter";
const LEGACY_PLIST = join(AGENTS, `${LEGACY_LABEL}.plist`);

const domain = () => `gui/${process.getuid!()}`;
const xml = (s: string) => s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");

/** Path of the running .app bundle, or null when not running from one. */
export function appBundlePath(from = process.execPath): string | null {
  const i = from.lastIndexOf(".app/");
  return i < 0 ? null : from.slice(0, i + 4);
}

export function isLaunchAtLogin(): boolean {
  return existsSync(LOGIN_PLIST);
}

/** A LaunchAgent that just opens the app at login (in the background, no KeepAlive). */
export function setLaunchAtLogin(enabled: boolean, appPath = appBundlePath()) {
  if (!enabled) {
    if (existsSync(LOGIN_PLIST)) unlinkSync(LOGIN_PLIST);
    return;
  }
  if (!appPath) throw new Error("Launch at login needs the app to run from its .app bundle");
  mkdirSync(dirname(LOGIN_PLIST), { recursive: true });
  writeFileSync(
    LOGIN_PLIST,
    `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key><string>${LOGIN_LABEL}</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/bin/open</string>
    <string>-g</string>
    <string>-a</string>
    <string>${xml(appPath)}</string>
  </array>
  <key>RunAtLoad</key><true/>
</dict>
</plist>
`,
  );
}

/** Stop and remove the old CLI daemon so it doesn't run checks alongside the app. */
export function removeLegacyDaemon(): boolean {
  if (!existsSync(LEGACY_PLIST)) return false;
  Bun.spawnSync(["launchctl", "bootout", `${domain()}/${LEGACY_LABEL}`]);
  unlinkSync(LEGACY_PLIST);
  return true;
}

File: src/go/core/loginitem.go
package core

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const loginLabel = "com.usage-window-starter.login"

// legacyLabel is the launchd daemon from the CLI-only version; the menu bar app replaces it.
const legacyLabel = "com.limit-starter"

func agentsDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, "Library/LaunchAgents")
}

func loginPlist() string { return filepath.Join(agentsDir(), loginLabel+".plist") }

// oldLoginPlist is the login item from before the rename; it opens the old .app.
func oldLoginPlist() string { return filepath.Join(agentsDir(), "com.limit-starter.login.plist") }

func legacyPlist() string { return filepath.Join(agentsDir(), legacyLabel+".plist") }

// AppBundlePath returns the running .app bundle, or "" when not running from one.
func AppBundlePath() string {
	exe, err := os.Executable()
	if err != nil {
		return ""
	}
	i := strings.LastIndex(exe, ".app/")
	if i < 0 {
		return ""
	}
	return exe[:i+4]
}

func IsLaunchAtLogin() bool { return exists(loginPlist()) }

var xmlEscape = strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;")

// SetLaunchAtLogin writes a LaunchAgent that just opens the app at login (in the
// background, no KeepAlive).
func SetLaunchAtLogin(enabled bool, appPath string) error {
	if !enabled {
		if err := os.Remove(loginPlist()); err != nil && !os.IsNotExist(err) {
			return err
		}
		return nil
	}
	if appPath == "" {
		return errors.New("Launch at login needs the app to run from its .app bundle")
	}
	if err := os.MkdirAll(agentsDir(), 0o755); err != nil {
		return err
	}
	return os.WriteFile(loginPlist(), []byte(fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key><string>%s</string>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/bin/open</string>
    <string>-g</string>
    <string>-a</string>
    <string>%s</string>
  </array>
  <key>RunAtLoad</key><true/>
</dict>
</plist>
`, loginLabel, xmlEscape.Replace(appPath))), 0o644)
}

// RemoveLegacyDaemon stops and removes the old CLI daemon so it doesn't run checks
// alongside the app.
func RemoveLegacyDaemon() bool {
	if !exists(legacyPlist()) {
		return false
	}
	_ = exec.Command("launchctl", "bootout", fmt.Sprintf("gui/%d/%s", os.Getuid(), legacyLabel)).Run()
	_ = os.Remove(legacyPlist())
	return true
}

// MigrateLoginItem swaps the pre-rename login item for one that opens this app.
func MigrateLoginItem(appPath string) bool {
	if !exists(oldLoginPlist()) {
		return false
	}
	_ = os.Remove(oldLoginPlist())
	if appPath != "" {
		_ = SetLaunchAtLogin(true, appPath)
	}
	return true
}
